python/example_code/sts/sts_temporary_credentials/assume_role_mfa.py

```diff
@@ -0,0 +1,247 @@
+
+"""
+Purpose
+
+Shows how to assume a role that requires a multi-factor authentication token,
+using AWS Security Token Service (STS) credentials.
+"""
+
+import json
+import os
+import time
+import sys
+import webbrowser
+import boto3
+from botocore.exceptions import ClientError
+
+
+def progress_bar(seconds):
+    """Shows a simple progress bar in the command window."""
+    for _ in range(seconds):
+        time.sleep(1)
+        print('.', end='')
+        sys.stdout.flush()
+    print()
+
+
+def unique_name(base_name):
+    return f'aws-assume-role-demo-{base_name}-{time.time_ns()}'
+
+
+def setup(iam_resource):
+    """
+    Creates a new user with no permissions.
+    Creates a new virtual multi-factor authentication (MFA) device.
+    Displays the QR code to seed the device.
+    Asks for two codes from the MFA device.
+    Registers the MFA device for the user.
+    Creates an access key pair for the user.
+    Creates a role with a policy that lets the user assume the role and requires MFA.
+    Creates a policy that allows listing Amazon S3 buckets.
+    Attaches the policy to the role.
+    Creates an inline policy for the user that lets the user assume the role.
+
+    For demonstration purposes, the user is created in the same account as the role,
+    but in practice the user would likely be from another account.
+
+    Any MFA device that can scan a QR code will work with this demonstration.
+    Common choices are mobile apps like LastPass Authenticator,
+    Microsoft Authenticator, or Google Authenticator.
+
+    :param iam_resource: A Boto3 AWS Identity and Access Management (IAM) resource
+                         that has permissions to create users, roles, and policies
+                         in the account.
+    :return: The newly created user, user key, virtual MFA device, and role.
+    """
+    user = iam_resource.create_user(UserName=unique_name('user'))
+    print(f"Created user {user.name}.")
+
+    virtual_mfa_device = iam_resource.create_virtual_mfa_device(
+        VirtualMFADeviceName=unique_name('mfa'))
+    print(f"Created virtual MFA device {virtual_mfa_device.serial_number}")
+
+    print(f"Showing the QR code for the device. Scan this in the MFA app of your "
+          f"choice.")
+    with open('qr.png', 'wb') as qr_file:
+        qr_file.write(virtual_mfa_device.qr_code_png)
+    webbrowser.open(qr_file.name)
+
+    print(f"Enter two consecutive code from your MFA device.")
+    mfa_code_1 = input("Enter the first code: ")
+    mfa_code_2 = input("Enter the second code: ")
+    user.enable_mfa(
+        SerialNumber=virtual_mfa_device.serial_number,
+        AuthenticationCode1=mfa_code_1,
+        AuthenticationCode2=mfa_code_2)
+    os.remove(qr_file.name)
+    print(f"MFA device is registered with the user.")
+
+    user_key = user.create_access_key_pair()
+    print(f"Created access key pair for user.")
+
+    print(f"Wait for user to be ready.", end='')
+    progress_bar(10)
```

Comment:
I don't recall off the top of my head if we have waiters for all of these things. I know that IAM can be a little tricky to nail down on when a particular resource is available so there may not be an applicable waiter that's reliable. If you've already looked into the waiters feel free to disregard this.

Reply:
Thanks for the thought! I did look into the waiters, and they work as advertised in that they indicate when the various individual pieces are ready (existence of user, role, policy), but in this case I need to wait until all of the pieces are wired up together behind the scenes (which seems to pretty consistently take around 10 seconds). Otherwise, I get an AccessDenied when trying to assume the role. I hunted around for a more deterministic way, but didn't come up with anything other than maybe calling get_account_authorization_details and examining the results, but that seemed pretty heavy-handed.